### DISCUSSION TITLE
Manipulate properties (values) by name (instead of expression)

### DISCUSSION THREAD

**Discussion by utillity** [Gremlinq/ExRam.Gremlinq#909] [General]
Hi Daniel,

I appreciate that ExRam.Gremlinq is a linq wrapper for gremlin and as such manipulating properties in a typed manner is the essence of what you're providing.

My need to be able to add, replace and remove values to properties in a non-typed/expression way is because I would like to automate updating DB entities from an internal state-object which only carries properties which need to be changed. I use reflection to find the corresponding property and value-type on the vertex and then I add the respective manipulation method to the query.

For "single" properties this is already supported by using `ExRam.Gremlinq.Core.IElementGremlinQueryBaseRec[TSelf].Property`. But for properties with a value-array the mentioned method only supports the `Add to list` aspect. For `Replace list` I use the private method `ExRam.Gremlinq.Core.GremlinQuery<TElement,TOutVertex,TInVertex,TScalar,TMeta,TFoldedQuery>.DropProperties`, but for `Remove from list` I found no method at all.

Would it make sense to make your private `DropProperties` public and provide a similar method ie. `DropPropertiesByValue`? Or do you suggest to go a different route?

thanks! Tilli

**Comment by utillity** [comment]
Any thoughts from your end on this?

**Comment by danielcweber** [comment]
I'm afraid I don't comprehend what you are proposing. What's the corresponding raw Gremlin query that you are trying to match in ExRam.Gremlinq?

**Comment by utillity** [comment]
Issue 1: removing a specific value from an array property:
Currently I can only drop a value from a design-time known property - ie.
`.SideEffect(_ => _.Properties(e => e.InternalTags).Where(p => values.Contains(p.Value)).Drop());`
Instead I would like to pass the property-name by string - ie.
`.SideEffect(_ => _.Properties("InternalTags").Where(p => values.Contains(p.Value)).Drop());`

Issue 2: replacing a property - you have an internal method `DropProperties` for this:
`// Query: g.V(["Test","8UQ9JT5D5TAJ0"]).hasLabel("TestVertex").sideEffect(__.properties("ListOne").drop()).property(list,"ListOne","V1").property(list,"ListOne","V2")`
`
// Code:  q = graph.V<TestVertex>(v.Key()).SideEffect(_ => _.Properties(e => e.ListOne).Drop()).Property(e => e.ListOne, new[] { "V1", "V2" });
`

**Comment by danielcweber** [comment]
I don't see a `Properties`-signature that would accept strings for reasons you mentioned above. Doing so would open pandora's box for all kinds of string-arguments everywhere and the original idea of Gremlinq would just suffer. What's the use case where you couldn't have a proper property on a POCO and reference this? What use case makes the property name that dynamic and unknown at compile time?

As for the second proposal: The behaviour you're observing is that list properties are added to the already existing ones, right? The problem with a potential "ReplaceProperty" method is that for non-list-properties, everything is a replace. So at this time I'd suggest creating an extension method in your project. I'd have to think more about this though.

**Comment by utillity** [comment]
Hi Daniel,

let me explain my use case a bit better: For each Vertex I have an Update-Request object which has the same properties as the Vertex, but all of them are nullable. To avoid having to write the if/then logic for each Property in each Vertex, I have created a reflection-based handler, which extends the query for each property that changes. It also logs the changes. 
This allows me to:
1) have consistent update handling and change-tracking
2) vastly reduce the code required
3) add update-logic for new vertices with minimal effort

I can change the value of the property by passing the property-name instead of an expression, but I cannot DROP the property, if the value has been set to NULL on the Update-Request.

**Comment by danielcweber** [comment]
Two questions:
>I can change the value of the property by passing the property-name instead of an expression

How? Can you post a snippet?

>but I cannot DROP the property

What happens if you pass null as updated value? It should work.
